backend/src/main/java/club/devcord/devmarkt/auth/UserProvider.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2022 Contributors to the Devmarkt-Backend project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package club.devcord.devmarkt.auth;
+
+import club.devcord.devmarkt.entities.auth.Role;
+import club.devcord.devmarkt.entities.auth.User;
+import club.devcord.devmarkt.services.UserService;
+import jakarta.inject.Singleton;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+@Singleton
+public class UserProvider {
+
+  private static final Pattern TOKEN_PATTERN = Pattern.compile("(Self|Foreign) .+");
+
+  private final UserIdValidator userIdValidator;
+  private final UserService userService;
+
+  public UserProvider(UserIdValidator userIdValidator,
+      UserService userService) {
+    this.userIdValidator = userIdValidator;
+    this.userService = userService;
+  }
+
+  public Optional<User> validate(String token) {
+    if (!TOKEN_PATTERN.matcher(token).matches()) {
+      return Optional.empty();
+    }
+
+    var split = token.split(" ", 2);
+    return switch (split[0]) {
+      case "Self" -> validateSelf(split[1]);
+      case "Foreign" -> validateForeign(split[1]);
+      default -> Optional.empty();
+    };
+  }
+
+  private Optional<User> validateSelf(String token) {
+    var id = userIdValidator.validateUserIdFromToken(token);
+    return id != null
+        ? userService.findDirect(id)
+        : Optional.empty();
+  }
+
+  private Optional<User> validateForeign(String value) {
+    var split = value.split(" ", 2);
```

Comment:
There's no need for more than 2 sub strings here and since the UserIdValidator does its own checks, I think this is the best way to validate and parse this. But actually there should be a null check after line 68.